src/security-archive/src/index.ts

```diff
@@ -0,0 +1,297 @@
+ */
+const usesTargetRegistry = (
+  depID: DepID,
+  specOptions: SpecOptions,
+): boolean => {
+  const [nodeType, nodeRegistry] = splitDepID(depID)
+
+  if (nodeType !== 'registry') {
+    return false
+  }
+
+  const reg =
+    nodeRegistry === '' ?
+      specOptions.registry
+    : specOptions.registries?.[nodeRegistry]
+  return reg === targetSecurityRegisty
+}
+
+// Loads the version number to be used in the User-Agent header
+export const { version } = loadPackageJson(
+  import.meta.filename,
+  process.env.__VLT_INTERNAL_CLI_PACKAGE_JSON,
+) as {
+  version: string
+}
+
+/**
+ * A database of security information for given packages in a graph.
+ *
+ * Using the SecurityArchive.refresh() method will update the local cache
+ * with information from the socket.dev APIs or load from the local storage
+ * if available. Information about package security is then available
+ * using the SecurityArchive.get() method.
+ */
+export class SecurityArchive
+  extends LRUCache<DepID, PackageReportData>
+  implements SecurityArchiveLike
+{
+  #path: string
+
+  /**
+   * Creates a new security archive instance and starts the refresh process.
+   */
+  static async start(
+    options: SecurityArchiveOptions & SecurityArchiveRefreshOptions,
+  ) {
+    const archive = new SecurityArchive(options)
+    await archive.refresh(options)
+    return archive
+  }
+
+  /**
+   * By default, caches up to 100K entries in the in-memory archive.
+   */
+  static get defaultMax() {
+    return 100_000
+  }
+
+  /**
+   * By default, entries are cached for 3 hour.
+   */
+  static get defaultTtl() {
+    return 1000 * 60 * 60 * 3
+  }
+
+  constructor(options: SecurityArchiveOptions = {}) {
+    super({
+      max: SecurityArchive.defaultMax,
+      ttl: SecurityArchive.defaultTtl,
+    })
+
+    this.#path =
+      options.path ?? new XDG('vlt').cache('security-archive.db')
+  }
+
+  /**
+   * Opens a connection to the database at the given path.
+   */
+  #openDatabase(): DatabaseSync {
+    const db = new DatabaseSync(this.#path)
+    db.exec(
+      'CREATE TABLE IF NOT EXISTS cache (key TEXT PRIMARY KEY, value TEXT)',
+    )
+    db.exec('PRAGMA synchronous = NORMAL')
+    return db
+  }
+
+  /**
+   * Loads all valid items found in the database into the in-memory cache.
+   */
+  #loadItemsFromDatabase(db: DatabaseSync, graph: GraphLike): void {
+    let keys = ''
+    for (const node of graph.nodes.values()) {
+      keys += `'${node.id}',`
+    }
+    keys = keys.substring(0, keys.length - 1)
+    const dbRead = db.prepare(
+      `SELECT key, value FROM cache WHERE key IN (${keys})`,
+    )
+    for (const entry of dbRead.all() as {
+      key: string
+      value: string
+    }[]) {
+      const { start, ttl, value } = JSON.parse(
+        entry.value,
+      ) as LoadedEntry
+      this.set(asDepID(entry.key), asPackageReportData(value), {
+        ttl,
+        start,
+      })
+    }
+    this.purgeStale()
```

Comment:
Why is this needed? Can we avoid it by just not adding anything that we know is past its TTL?

Reply:
how could we avoid not adding anything that we know is past TTL? I think that's a better solution 👍